# Intro to Pandas (5) - Accessing and Changing Specific Observations

In the last lesson we saw how to rename and drop columns, and to set the index in a DataFrame.

In this lesson we'll learn about positional and label-based selection and how to use this to make changes to specific observations.

## Module Imports

In [1]:
#plotly.offline doesn't push your charts to the clouds
import plotly.offline as pyo
#allows us to create the Data and Figure objects
from plotly.graph_objs import *
#plotly.plotly pushes your charts to the cloud  
import plotly.plotly as py

#pandas is a data analysis library
import pandas as pd
from pandas import DataFrame

## Positional and Label-based Selection

First of all, I'm going to read some data into a DataFrame. I also need to make another DataFrame which has a label-based index rather than a positional index.

In [2]:
baseRateData = pd.read_csv("http://www.richard-muir.com/data/public/csv/BoEBaseRate.csv")

baseRateData_r = baseRateData.rename(columns = {'VALUE' : 'Value', 'DATE' : 'Date'})
baseRateData_r.set_index(baseRateData_r['Date'], inplace=True)
baseRateData_r.drop(['Date'], axis = 1, inplace = True)

Let's have a look at these DataFrames:

In [3]:
baseRateData.head()

VALUE        DATE
0   11.5  02/01/1975
1   11.5  03/01/1975
2   11.5  06/01/1975
3   11.5  07/01/1975
4   11.5  08/01/1975

In [4]:
baseRateData_r.head()

Value
Date             
02/01/1975   11.5
03/01/1975   11.5
06/01/1975   11.5
07/01/1975   11.5
08/01/1975   11.5

#### Selecting observations in a DataFrame

We can select observations from a DataFrame by using <code>df.loc</code> and <code>df.iloc</code>.

- <code>df.loc</code> selects the observations by their index label
- <code>df.iloc</code> selects the observations by their position

Here, I'm using <code>df.loc</code> to select the first 10 rows from <code>baseRateData</code>. Note that <code>df.loc</code> doesn't work like a list slice in Python; rather than stopping before the specified number, we include that observation:

In [5]:
baseRateData.loc[:9]

VALUE        DATE
0   11.5  02/01/1975
1   11.5  03/01/1975
2   11.5  06/01/1975
3   11.5  07/01/1975
4   11.5  08/01/1975
5   11.5  09/01/1975
6   11.5  10/01/1975
7   11.5  13/01/1975
8   11.5  14/01/1975
9   11.5  15/01/1975

If I try to use <code>df.loc</code> on <code>baseRateData_r</code>, this won't work because we have changed the index label:

In [6]:
baseRateData_r.loc[:9]

TypeError: cannot do slice indexing on <class 'pandas.indexes.base.Index'> with these indexers [9] of <class 'int'>

Instead I have to pass the row index label which I want:

In [7]:
baseRateData_r.loc[:'15/01/1975']

Value
Date             
02/01/1975   11.5
03/01/1975   11.5
06/01/1975   11.5
07/01/1975   11.5
08/01/1975   11.5
09/01/1975   11.5
10/01/1975   11.5
13/01/1975   11.5
14/01/1975   11.5
15/01/1975   11.5

But <code>df.iloc</code> works the same on both DataFrames because in <code>baseRateData</code>, the index is equal to the position - <code>df.iloc</code> works on the ordinal position of the rows.

Confusingly, <code>df.iloc</code> works in the same way as list and string slicing, stopping just before the specified position:

In [8]:
baseRateData.iloc[:9]

VALUE        DATE
0   11.5  02/01/1975
1   11.5  03/01/1975
2   11.5  06/01/1975
3   11.5  07/01/1975
4   11.5  08/01/1975
5   11.5  09/01/1975
6   11.5  10/01/1975
7   11.5  13/01/1975
8   11.5  14/01/1975

In [9]:
baseRateData_r.iloc[:9]

Value
Date             
02/01/1975   11.5
03/01/1975   11.5
06/01/1975   11.5
07/01/1975   11.5
08/01/1975   11.5
09/01/1975   11.5
10/01/1975   11.5
13/01/1975   11.5
14/01/1975   11.5

For both <code>df.loc</code> and <code>df.iloc</code>, we can take a slice from the middle of the DataFrame:

In [10]:
baseRateData_r.loc['06/01/1975':'13/01/1975']

Value
Date             
06/01/1975   11.5
07/01/1975   11.5
08/01/1975   11.5
09/01/1975   11.5
10/01/1975   11.5
13/01/1975   11.5

In [11]:
baseRateData.iloc[4:6]

VALUE        DATE
4   11.5  08/01/1975
5   11.5  09/01/1975

We can also combine the column names with <code>df.loc</code> and <code>df.iloc</code> to get 2D slices of a DataFrame.

Remember that <code>df.loc</code> works on the labels:

In [12]:
baseRateData.loc[5:13, 'DATE']

5     09/01/1975
6     10/01/1975
7     13/01/1975
8     14/01/1975
9     15/01/1975
10    16/01/1975
11    17/01/1975
12    20/01/1975
13    21/01/1975
Name: DATE, dtype: object

But <code>df.iloc</code> operates on the index; the columns are numerically indexed (in the same way as the rows):

In [13]:
baseRateData.iloc[5:13, 0]

5     11.50
6     11.50
7     11.50
8     11.50
9     11.50
10    11.50
11    11.50
12    11.25
Name: VALUE, dtype: float64

### Changing Data in a DataFrame

So now we can select individual rows and columns in a DataFrame by the index label or position. We can use this knowledge to make changes to specific observations within the DataFrame.

Imagine that we were told that the first twenty rows of our data were incorrect; they should have been 1.15 instead of 11.5. Let's make some changes!

First of all, I'm using <code>df.loc</code> to select the first 20 rows by label and only the 'VALUE' column. It's just a simple matter of setting the value which we want these observations to take:

In [14]:
baseRateData.loc[:19, 'VALUE'] = 1.15

baseRateData.head(25)

VALUE        DATE
0    1.15  02/01/1975
1    1.15  03/01/1975
2    1.15  06/01/1975
3    1.15  07/01/1975
4    1.15  08/01/1975
5    1.15  09/01/1975
6    1.15  10/01/1975
7    1.15  13/01/1975
8    1.15  14/01/1975
9    1.15  15/01/1975
10   1.15  16/01/1975
11   1.15  17/01/1975
12   1.15  20/01/1975
13   1.15  21/01/1975
14   1.15  22/01/1975
15   1.15  23/01/1975
16   1.15  24/01/1975
17   1.15  27/01/1975
18   1.15  28/01/1975
19   1.15  29/01/1975
20  11.00  30/01/1975
21  11.00  31/01/1975
22  11.00  03/02/1975
23  11.00  04/02/1975
24  11.00  05/02/1975

We can also do it with <code>df.iloc</code>. Remember that the slicing is slightly different...

I'll change it instead to 2.15 so we can prove it works:

In [15]:
baseRateData.iloc[:20, 0] = 2.15

baseRateData.head(25)

VALUE        DATE
0    2.15  02/01/1975
1    2.15  03/01/1975
2    2.15  06/01/1975
3    2.15  07/01/1975
4    2.15  08/01/1975
5    2.15  09/01/1975
6    2.15  10/01/1975
7    2.15  13/01/1975
8    2.15  14/01/1975
9    2.15  15/01/1975
10   2.15  16/01/1975
11   2.15  17/01/1975
12   2.15  20/01/1975
13   2.15  21/01/1975
14   2.15  22/01/1975
15   2.15  23/01/1975
16   2.15  24/01/1975
17   2.15  27/01/1975
18   2.15  28/01/1975
19   2.15  29/01/1975
20  11.00  30/01/1975
21  11.00  31/01/1975
22  11.00  03/02/1975
23  11.00  04/02/1975
24  11.00  05/02/1975

### What have we learnt this lesson?

In this lesson we've seen how to access rows and columns by their label and position, and to use this positional selection to make changes to the data in the DataFrame.

If you have any questions, please ask in the comments section or email <a href="mailto:me@richard-muir.com">me@richard-muir.com</a>